### PR TITLE
Ported get_keymap C function

### LINKED ISSUE #668: Port get_keymap from keymap.c
Eyeing to work on this next one from `keymap.c`.

```c
/* Check that OBJECT is a keymap (after dereferencing through any
   symbols).  If it is, return it.

   If AUTOLOAD and if OBJECT is a symbol whose function value
   is an autoload form, do the autoload and try again.
   If AUTOLOAD, callers must assume GC is possible.

   ERROR_IF_NOT_KEYMAP controls how we respond if OBJECT isn't a keymap.
   If ERROR_IF_NOT_KEYMAP, signal an error; otherwise,
   just return Qnil.

   Note that most of the time, we don't want to pursue autoloads.
   Functions like Faccessible_keymaps which scan entire keymap trees
   shouldn't load every autoloaded keymap.  I'm not sure about this,
   but it seems to me that only read_key_sequence, Flookup_key, and
   Fdefine_key should cause keymaps to be autoloaded.

   This function can GC when AUTOLOAD is true, because it calls
   Fautoload_do_load which can GC.  */

Lisp_Object
get_keymap (Lisp_Object object, bool error_if_not_keymap, bool autoload)
{
  Lisp_Object tem;

 autoload_retry:
  if (NILP (object))
    goto end;
  if (CONSP (object) && EQ (XCAR (object), Qkeymap))
    return object;

  tem = indirect_function (object);
  if (CONSP (tem))
    {
      if (EQ (XCAR (tem), Qkeymap))
	return tem;

      /* Should we do an autoload?  Autoload forms for keymaps have
	 Qkeymap as their fifth element.  */
      if ((autoload || !error_if_not_keymap) && EQ (XCAR (tem), Qautoload)
	  && SYMBOLP (object))
	{
	  Lisp_Object tail;

	  tail = Fnth (make_number (4), tem);
	  if (EQ (tail, Qkeymap))
	    {
	      if (autoload)
		{
		  Fautoload_do_load (tem, object, Qnil);
		  goto autoload_retry;
		}
	      else
	      	return object;
	    }
	}
    }

 end:
  if (error_if_not_keymap)
    wrong_type_argument (Qkeymapp, object);
  return Qnil;
}
```

This particular function uses `goto` statements so I reckon I'll be doing some work with restructuring the logic for this.

### PR DESCRIPTION
**Pull request by accidentalrebel**
Ported the `get_keymap` C function. Closes #668.

Notes:
- Other parts of `keymap.rs` needed to be updated by removing the now unneeded `unsafe` blocks.
- I had to also update `data.rs` which also uses `get_keymap`.
- Tested working by making temporary lisp tests.
- Initially used Rust labels but refactored it so there are no more nested loops.

### CODE REVIEW

File: rust_src/src/keymap.rs [hunk 1]
```diff
@@ -51,6 +53,90 @@ pub extern "C" fn set_where_is_cache_keymaps(val: Lisp_Object) {
...
+        if object.is_nil() {
+            break;
+        }
+
+        if object.is_cons()
+            && object
+                .as_cons()
+                .unwrap()
+                .car()
+                .eq(LispObject::from_raw(Qkeymap))
```
Comment (shaleh) [inline rust_src/src/keymap.rs]: if let Some(cons) = object.as_cons() {
        if cons.car().eq_raw(Qkeymap) {
            return object.to_raw();
        }
    }

File: rust_src/src/keymap.rs [hunk 2]
```diff
@@ -51,6 +53,90 @@ pub extern "C" fn set_where_is_cache_keymaps(val: Lisp_Object) {
...
+        {
+            return object.to_raw();
+        }
+
+        let tem = indirect_function(object);
+        if tem.is_cons() {
+            if tem.as_cons()
+                .unwrap()
+                .car()
+                .eq(LispObject::from_raw(Qkeymap))
```
Comment (shaleh) [inline rust_src/src/keymap.rs]: Same as above.

File: rust_src/src/keymap.rs [hunk 3]
```diff
@@ -51,6 +53,90 @@ pub extern "C" fn set_where_is_cache_keymaps(val: Lisp_Object) {
...
+
+            // Should we do an autoload?  Autoload forms for keymaps have
+            // Qkeymap as their fifth element.
+            if (autoload || !error_if_not_keymap)
+                && tem.as_cons()
+                    .unwrap()
+                    .car()
+                    .eq(LispObject::from_raw(Qautoload)) && object.is_symbol()
+            {
+                let tail = nth(4, tem);
```
Comment (shaleh) [inline rust_src/src/keymap.rs]: Again with `eq_raw`

File: rust_src/src/keymap.rs [hunk 4]
```diff
@@ -51,6 +53,90 @@ pub extern "C" fn set_where_is_cache_keymaps(val: Lisp_Object) {
...
+/// `Fautoload_do_load` which can GC.
+#[no_mangle]
+pub extern "C" fn get_keymap(
+    object: Lisp_Object,
+    error_if_not_keymap: bool,
+    autoload: bool,
+) -> Lisp_Object {
+    let object = LispObject::from_raw(object);
+    let mut autoload_retry;
+    loop {
```
Comment (shaleh) [inline rust_src/src/keymap.rs]: This is a while loop based on the `autoload_rety` variable.

Comment (accidentalrebel) [reply rust_src/src/keymap.rs]: I initially wanted to use a `do while` loop but learned that Rust doesn't have one. Switched to this one. But I agree maybe a `while` loop is better.

File: rust_src/src/keymap.rs (hunk 3 above)
Comment (accidentalrebel) [reply rust_src/src/keymap.rs]: This is quite useful! Thanks!

File: rust_src/src/keymap.rs (hunk 2 above)
Comment (accidentalrebel) [reply rust_src/src/keymap.rs]: Got it.

File: rust_src/src/keymap.rs (hunk 1 above)
Comment (accidentalrebel) [reply rust_src/src/keymap.rs]: Gotcha. The long chain should have been my clue to break it down.

**Comment by accidentalrebel** [comment]
Alright. Fixed the requested changes. :)

**Comment by Wilfred** [comment]
@accidentalrebel I've added you as a collaborator, so you now have the power to approve PRs as well if you fancy :)

**Pull request merged**